src/JsonReader.cs

```diff
@@ -548,127 +570,162 @@ sealed class NonProperty : IJsonProperty<Unit, JsonReadResult<Unit>>
             IJsonProperty<T13, JsonReadResult<T13>> property13, IJsonProperty<T14, JsonReadResult<T14>> property14,
             IJsonProperty<T15, JsonReadResult<T15>> property15, IJsonProperty<T16, JsonReadResult<T16>> property16,
             Func<T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15, T16, TResult> projector) =>
-        CreatePure((ref Utf8JsonReader reader) =>
+        Create((ref Utf8JsonReader reader) =>
         {
-            if (reader.TokenType is JsonTokenType.None && !reader.Read())
-                throw PartialJsonNotSupportedException();
-
-            var tokenType = reader.TokenType;
-            if (tokenType is not JsonTokenType.StartObject)
-                return Error("Invalid JSON value where a JSON object was expected.");
+            var (sm, values) =
+                reader.IsResuming && ((ObjectReadStateMachine, ObjectValueState<T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15, T16>))reader.Pop() is var ps
+                    ? ps
+                    : default;
 
-            (bool, T1) value1 = default;
-            (bool, T2) value2 = default;
-            (bool, T3) value3 = default;
-            (bool, T4) value4 = default;
-            (bool, T5) value5 = default;
-            (bool, T6) value6 = default;
-            (bool, T7) value7 = default;
-            (bool, T8) value8 = default;
-            (bool, T9) value9 = default;
-            (bool, T10) value10 = default;
-            (bool, T11) value11 = default;
-            (bool, T12) value12 = default;
-            (bool, T13) value13 = default;
-            (bool, T14) value14 = default;
-            (bool, T15) value15 = default;
-            (bool, T16) value16 = default;
+            return Read(ref reader, sm, ref values);
 
-            while (true)
+            JsonReadResult<TResult> Read(ref Utf8JsonReader reader, ObjectReadStateMachine sm, ref ObjectValueState<T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15, T16> values)
             {
-                string? error = null;
+                while (true)
+                {
+                    switch (sm.Read(ref reader))
+                    {
+                        case ObjectReadStateMachine.ReadResult.Error:
+                            return Error("Invalid JSON value where a JSON object was expected.");
 
-                if (!reader.TryReadToken(out tokenType))
-                    throw PartialJsonNotSupportedException();
+                        case ObjectReadStateMachine.ReadResult.Incomplete:
+                            return reader.Suspend((sm, values));
 
-                if (tokenType is JsonTokenType.EndObject)
-                    break;
+                        case ObjectReadStateMachine.ReadResult.Done:
```

Comment:
Some `case` arms are large so I would put them all consistently in a block (`{`…`}`) (also scopes the variables of each).

Reply:
nit: To aid in linear reading and comprehension, it would be best to make the `Done` case the last in the `switch`.

Reply:
Done in 7b8fc262061d129ba4df47354d9efb941ae61e22